server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -336,7 +344,25 @@ public final void addDynamicMapper(Mapper mapper) {
         // dynamically mapped objects when the incoming document defines no sub-fields in them:
         // 1) by default, they would be empty containers in the mappings, is it then important to map them?
         // 2) they can be the result of applying a dynamic template which may define sub-fields or set dynamic, enabled or subobjects.
-        dynamicMappers.computeIfAbsent(mapper.name(), k -> new ArrayList<>()).add(mapper);
+        dynamicMappers.computeIfAbsent(mapper.name(), k -> new ArrayList<>(1)).add(mapper);
+        return true;
+    }
+
+    /*
+     * Returns an approximation of the number of dynamically mapped fields that will be added to the mapping.
+     * This is to validate early and to fail fast during document parsing.
+     * There will be another authoritative (but more expensive) validation step when making the actual update mapping request.
+     * During the mapping update, the actual number fields is determined by counting the total number of fields of the merged mapping.
+     * Therefore, both over-counting and under-counting here is not critical.
+     * However, in order for users to get to the field limit, we should try to be as close as possible to the actual field count.
+     * If we under-count fields here (for example by not counting multi-fields),
+     * we may only know that we exceed the field limit during the mapping update.
+     * This leads to document rejection instead of ignoring fields above the limit if Dynamic.UNTIL_LIMIT is configured for the index.
+     * If we over-count the fields (for example by counting all mappers with the same name),
+     * we may reject fields earlier than necessary and before actually hitting the field limit.
+     */
+    private int getNewDynamicMappersSize() {
```

Comment:
I feel like this method is tested enough implicitly via the new tests in `DynamicMappingIT`. There are no unit tests for `DocumentParserContext` yet so it would require a new testing setup as well as making the `getNewDynamicMappersSize()` method package-private instead of private. IMHO the effort/benefit ratio doesn't seem worth it. If we want to test this method more, I feel like we can do that by adding more tests to `DynamicMappingIT`.

 If you still want unit tests here, I'd prefer doing that as a follow-up.

Reply:
I do think that we should add unit tests and they are very different from DynamicMappingIT which is an integration test. Like I said, it can be done as a follow-up. And making methods package private for testing is perfectly fine.